### PR TITLE
[2201.2.x] Add fill record fields code action

### LINKED ISSUE #37647: [New Feature]: Add code action to auto fill a map based on the type
### Description

Right now when user is trying to construct an object we get auto fill as a suggestion (when we press `ctrl + space`) which will fill out all the required fields. 
we also get the diagnostics as given below. 
It would be better if we can get a code action to auto fill all the required fields (similar to the suggestion) as it would be very useful for the low-code users.

quickFix

### Describe your problem(s)

see description

### Describe your solution(s)

_No response_

### Labels

LanguageServer

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by mindula**
## Purpose
This PR introduces a new code action to fill record fields.

Fixes #37647

## Samples

https://user-images.githubusercontent.com/61020198/193212380-553d4a62-dd81-49aa-b27c-e76ac78e662a.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by LakshanWeerasinghe — COMMENTED** [review]
Add a test case to check the code action is working for records with type reference.

```
type RecordName record {
    *RecordName2 rec1f1;
    int rec1f2;
};

type RecordName2 record {
    int rec2f1;
    int rec2f2;
};

RecordName recordName = {<cursor>}
```

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/NameUtil.java [hunk 1]
```diff
@@ -214,6 +215,24 @@ public static String getModifiedTypeName(DocumentServiceContext context, TypeSym
...
     }
 
+    /**
+     * Returns the record type.
+     *
+     * @param context Context
+     * @param wrapper A wrapper containing record type symbol and broader type symbol
+     * @return Record type
+     */
+    public static String getRecordType(DocumentServiceContext context,
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/NameUtil.java:225]: Maybe we can change this to `getRecordTypeName` WDYT? NameUtil javadoc - `Carries a set of utilities used for operations on names of variables, types, parameters, visible symbols.`

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/FillRecordFieldsCodeActionTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public String getResourceDir() {
+        return "fill-record-fields";
+    }
+
+    @DataProvider(name = "codeaction-data-provider")
+    @Override
+    public Object[][] dataProvider() {
+        return new Object[][] {
+                {"fill_record_fields_config1.json"},
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/FillRecordFieldsCodeActionTest.java:45]: Better if we can add descriptions to the config files, if we don't have a self explanatory file name.

**Review by gayalkuruppu — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/FillRecordFieldsCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        }
+        Optional<Node> mappingConstructorNode = getMappingConstructorNode(evalNode.get());
+        if (mappingConstructorNode.isEmpty()) {
+            return Collections.emptyList();
+        }
+        CodeActionContextTypeResolver contextTypeResolver = new CodeActionContextTypeResolver(context);
+        Optional<TypeSymbol> typeSymbol = mappingConstructorNode.get().apply(contextTypeResolver);
+        if (typeSymbol == null || typeSymbol.isEmpty()) {
+            return Collections.emptyList();
+        }
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/FillRecordFieldsCodeAction.java:95]: Will check again.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 4]
```diff
@@ -622,4 +622,21 @@ public static io.ballerina.tools.text.TextEdit getTextEdit(SyntaxTree syntaxTree
         int end = textDocument.textPositionFrom(LinePosition.from(endPos.getLine(), endPos.getCharacter()));
         return io.ballerina.tools.text.TextEdit.from(TextRange.from(start, end - start), textEdit.getNewText());
     }
+
+    /**
+     * Returns the matching node for a given node.
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:627]: Any suggestions?

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java [hunk 5]
```diff
@@ -622,4 +622,21 @@ public static io.ballerina.tools.text.TextEdit getTextEdit(SyntaxTree syntaxTree
...
+     *
+     * @param nodeAtCursor node
+     * @return node
+     */
+    public static Optional<Node> getEvalNode(Node nodeAtCursor) {
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:632]: This method was not added by me, i moved it to commonUtil for reuse purposes. So wont be changing the name.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/FillRecordFieldsCodeAction.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        return Optional.of(node);
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    static class ExtendsCodeActionNodeValidator extends CodeActionNodeValidator {
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/FillRecordFieldsCodeAction.java:150]: +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/FillRecordFieldsCodeAction.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+        @Override
+        public Boolean transform(SpecificFieldNode node) {
+            return super.transform(node) || (node.colon().isEmpty() && node.fieldName().toSourceCode().isEmpty()
+                    && node.valueExpr().isEmpty()) || (!node.fieldName().toSourceCode().isEmpty() &&
+                    node.colon().isEmpty());
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/FillRecordFieldsCodeAction.java:156]: +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/NameUtil.java (hunk 1 above)
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/NameUtil.java:225]: getRecordTypeName is better. i'll change it

**Review by mohanvive — APPROVED** [review]

**Review by TharushiJay — APPROVED** [review]
LGTM

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/NameUtil.java [hunk 8]
```diff
@@ -214,6 +215,24 @@ public static String getModifiedTypeName(DocumentServiceContext context, TypeSym
         return CommonUtil.getModifiedSignature(context, typeSignature);
     }
 
+    /**
+     * Returns the record type.
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/NameUtil.java:219]: `Returns record type name`?

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/FillRecordFieldsCodeActionTest.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import org.ballerinalang.langserver.commons.workspace.WorkspaceDocumentException;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+
+/**
+ * Test Cases for Fill Record Fields code action.
+ *
+ * @since 2201.2.2
```
Comment (IMS94) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/codeaction/FillRecordFieldsCodeActionTest.java:27]: 2201.2.3

**Review by IMS94 — CHANGES_REQUESTED** [review]
Well done with the implementation. I have added some comments to refine the logic mainly. Please address them

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/common/utils/CommonUtil.java:632]: +1

**Review by IMS94 — APPROVED** [review]

**Review by mohanvive — APPROVED** [review]

**Pull request merged**